Motor Zaber Motion

colbysparks opened the pull request:
Review for zaber motion's implementation of asyn motor controller and axis. This module makes use of ZaberMotionCppSupport package which sits in the `./zaberMotionSupport` folder. This package isn't included in the diff as it contains many shared libraries and header files.

This repo also includes scripts for updating the zml support package, and for configuring and building the zaber motor library, example ioc and unit tests.

Also, this github repository is a mirror of a zaber's public zaber-motion-epics gitlab repo.

colbysparks committed:
Connects controller and axis
- setting values of zaber motor record calls underlying functions now

colbysparks committed:
Stash working changes for zaber motor module

colbysparks committed:
Implements polling and addstry/catch func

colbysparks committed:
Rough outline of AXIS impl

colbysparks committed:
Adds clang format file

colbysparks committed:
Syncs changes with dev branch

colbysparks committed:
Resolve one last conflict

colbysparks committed:
Adds check for rotary / linear devices

colbysparks committed:
Rename performAction util function

colbysparks committed:
Starts work on zaber connection manager

colbysparks committed:
Implement report

colbysparks committed:
Cleans up print statements and adds flag checks

colbysparks committed:
Adds support module and starts work on updating makefile

colbysparks committed:
Updates makefile to build with support bundle

colbysparks committed:
Starts work on connection manager unit tests

colbysparks committed:
Merge branch 'dev' into review-branch

colbysparks committed:
Sync with changes on dev

colbysparks committed:
Adds unit tests for zaberConnectionManager

colbysparks committed:
Adds unit tests for removeConnection and RemovePrefix

colbysparks committed:
Updates fault-handling code -- Adds comments for status bits

colbysparks committed:
Add more comments for axis member functions

colbysparks committed:
Rename tests

colbysparks committed:
Adds example of connection with multiple devices

colbysparks committed:
Locks mutex for connection map access in connection manager

colbysparks committed:
Adds some additional tests

colbysparks committed:
Spacing

colbysparks committed:
code cleanup

colbysparks committed:
Change zml flag error to warning

colbysparks committed:
Adds script for updating ZML support package

colbysparks committed:
Adds means for configuring local device db

colbysparks committed:
Adds EPICS shell fn for registering db path

colbysparks committed:
Remove comment about ZABER_DEVICE_DB_PATH

colbysparks committed:
Update ZML version

colbysparks committed:
Fixes configuration for macOS build

colbysparks committed:
Merge branch 'review-branch' of gitlab.izaber.com:colby.sparks/zaber-motor-epics into review-branch

colbysparks committed:
Changes from merge requst
- hard-code s3 bucket url in gulpfile.js
- add error return to device db path setter ioc shell function
- update report functions to use device / axis toString
- fix spacing in zaberUtils

colbysparks committed:
Fix gulpfile formatting

colbysparks committed:
Adds more general way of handling faults

colbysparks committed:
Updates from merge request
- Suppress weak_ptr lock warning in connection manager
- Print error on connection failure in zaberController
- Improve unordered_map access in zaberAxis

colbysparks committed:
Adds documentation -- formatting elsewhere

colbysparks committed:
Cleanup: Remove motor status comments from zaberAxis

colbysparks committed:
Adds setup instructions

colbysparks committed:
Fix asyn and sequencer repo links in SETUP doc

colbysparks committed:
remove .vscode folder

colbysparks committed:
Adds docker image; Updates gulpfile for CI

colbysparks committed:
Update gulpfile for ubuntu build

colbysparks committed:
Update gulpfile build script
- Adds motorZaberMotion module to motor Makefile

colbysparks committed:
Remove comment and rename function in gulpfile

colbysparks committed:
Updates testing; Adds gitlab ci script
- Adds test function to gulpfile script
- Updates test/CMakeLists.txt to use ctest
- Fixes broken test in zaberConnectionManagerTest.cpp

colbysparks committed:
Update SETUP documentation

colbysparks committed:
adds comment on polling status updates to zaberAxis::poll

colbysparks committed:
Merge branch 'dev' into review-branch

colbysparks committed:
Update example ioc controller instantiation

kmpeters commented on iocs/zaberMotionIOC/iocBoot/iocZaberMotion/motor.cmd.zaber:
Are there really three controllers all sharing /dev/USB0?  Controllers from other vendors usually have a single USB connection per controller, so it is much more common to see USB0, USB1 and USB2 for a configuration like this.  Also, for systems that have multiple USB devices connection, it is much more convenient to use the soft-links in /dev/serial/by-id/ (on RHEL systems), rather than the /dev/ttyUSB# devices, because the by-id links don't depend on the order the devices were connected to the system.

colbysparks replied:
Zaber devices can be chained together on a single serial connection, so this is possible. Each device has a unique device chain index, which is passed as the last parameter of the create function.

colbysparks replied:
> Also, for systems that have multiple USB devices connection, it is much more convenient to use the soft-links in /dev/serial/by-id/ (on RHEL systems), rather than the /dev/ttyUSB# devices, because the by-id links don't depend on the order the devices were connected to the system.

I realized that I didn't respond to this portion of your comment. I'll update this to use soft-link and also throw in an example tcp configuration in case it might be useful.

colbysparks replied:
I added an example of how to set up a 2-axis controller with a tcp connection, but decided against using soft-links for serial because `/dev/ttyUSBX` seems salient enough (for me at least).

kmpeters commented on zaberMotionSupport/configure/RELEASE:
The ZML_HOST_ARCH, ZML_SUPPORT_LIB, and ZML_SUPPORT_INCLUDE should be moved to the CONFIG_SITE file, rather than reside in the RELEASE file.  My understanding is that the only thing that should reside in RELEASE files is module paths (or the inclusion of other files that contain module paths).  It is likely that the inclusion of these definitions in the RELEASE file has resulted in them also being included in the IOC's envPaths file, which doesn't make a lot of sense.

kmpeters replied:
Does the zaberMotionSupport directory need its own configure subdirectory?  I would have expected the build to work if the contents in the `zaberMotionSupport/RELEASE` file instead resided in the `configure/RELEASE` and `configure/CONFIG_SITE` files at the top of the module.

colbysparks replied:
So this should be moved to the top level config site? I was honestly a bit confused about how to define and include `ZML_SUPPORT_INCLUDE` and `ZML_SUPPORT_LIB` so I can access them from the app and ioc Makefile.

I agree that it doesn't actually make very much sense to define them in `zaberMotionSupport/configure` though

kmpeters commented on zaberMotionApp/src/Makefile:
It is probably a good idea to add _WIN32 or _Linux suffixes to the LIBRARY_IOC variable, like Mark Rivers recently did for the motorAutomation1 support, to avoid build problems when cross compiling for unsupported platfors like VxWorks or RTEMS in addition to building for host architectures like Linux:

https://github.com/epics-motor/motorAutomation1/pull/35/files#diff-8c5311d1a1d8f4d777d776f0396bb31d56cd632eb0f9afb20e8a2d0a0ba1fa42L12-R13

colbysparks replied:
Is there a suffix for macOS as well? I imagine it's not a very common use case, but I have been developing on mac

kmpeters replied:
It looks like **_Darwin** is the suffix for macOS:

https://github.com/areaDetector/ADCore/blob/0ed979a1c8e1f74a16d5c3d80d847417c582f42d/ADApp/pluginTests/Makefile#L27

kmpeters commented on SETUP.md:
This is much better documentation than I ususally write.  Users will appreciate it.

kmpeters commented on iocs/zaberMotionIOC/iocBoot/iocZaberMotion/st.cmd:
The approach used by the xxx (example synApps IOC) module is to provide separate st.cmd files for each support arch:

https://github.com/epics-modules/xxx/tree/master/iocBoot/iocxxx

It is a judgement call whether having lots of unused files or having to edit a file before it can be used is more annoying to users.

colbysparks replied:
If this is the precedent, then I think it makes sense to do it this way.

colbysparks committed:
Add separate ioc script for each supported sys/arch
Also update .gitignore to ignore .DS_STORE files

colbysparks commented on zaberMotionApp/src/zaberMotionRegister.cpp:
```diff
@@ -0,0 +1,73 @@
+{
+    zaberController *pController = new zaberController(portName, numAxes, static_cast<double>(movingPollPeriod), static_cast<double>(idlePollPeriod), zaberPort, zaberDeviceNumber);
+    (void)pController;
+    return (asynSuccess);
```
@kmpeters in the various motor repositories, it seems common for these `CreateController` iocsh functions to return `asynSuccess` or `asynError`. However from what I can tell, this returned value is always ignored as the call func which is actually registered with the epics IOC returns void, and doesn't do any kind of status handling. It's the same in the newport motor repo for instance:
https://github.com/epics-motor/motorNewport/blob/16a818fe7e49746ffd1973c47354631cb0793e81/newportApp/src/XPSController.cpp#L1344

I'm wondering why these functions return anything at all. Are they called asynchronously anywhere or only synchronously through the ioc shell?

kmpeters replied:
I think they return something because the return value could be checked, but in practice it never is.  I believe they are only called synchronously through the ioc shell.

colbysparks committed:
Update build scripting
- Adds system suffixes to LIBRARY_IOC vars
- Adds comment about compiling using -std=c++17
- Adds gulp function for building motor module

colbysparks commented:
@kmpeters I found out what is causing the `make clean` command to remove all files in `zaberMotionSupport`. The issue is that for whatever reason the make system is executing the target `build_clean` defined in `epics-base/configure/RULES_BUILD`:

```makefile
build_clean:
	$(ECHO) "Cleaning"
	$(RM) *.i *$(OBJ) *.a \
	$(LIBNAME) $(TESTLIBNAME) $(SHRLIBNAME) $(TESTSHRLIBNAME) \
	$(DLLSTUB_LIBNAME) $(TESTDLLSTUB_LIBNAME) \
	$(LOADABLE_SHRLIBNAME) \
	$(INC) $(TARGETS) $(TDS) $(CLEANS) \
	*.out MakefileInclude *.manifest *.exp \
	$(COMMON_INC) $(HDEPENDS_FILES) $(PRODTARGETS) $(TESTPRODTARGETS) \
	$(TESTSCRIPTS) $(TAPFILES) $(JUNITFILES)
ifdef RES
	@$(RM) *$(RES)
endif
```

The problem is that OBJ isn't defined and so it winds up running:
```bash
rm -f *.i * *.a *.out MakefileInclude *.manifest *.exp
```

`rm -f *` attempts to remove all files in a directory.

colbysparks commented on zaberMotionSupport/Makefile:
```diff
@@ -0,0 +1,18 @@
+
+#=============================
+
+include $(TOP)/configure/RULES
```
I found out what is causing the `make clean` command to remove all files in `zaberMotionSupport`. The issue is that for whatever reason the make system is executing the target `build_clean` defined in `epics-base/configure/RULES_BUILD`:

```makefile
build_clean:
	$(ECHO) "Cleaning"
	$(RM) *.i *$(OBJ) *.a \
	$(LIBNAME) $(TESTLIBNAME) $(SHRLIBNAME) $(TESTSHRLIBNAME) \
	$(DLLSTUB_LIBNAME) $(TESTDLLSTUB_LIBNAME) \
	$(LOADABLE_SHRLIBNAME) \
	$(INC) $(TARGETS) $(TDS) $(CLEANS) \
	*.out MakefileInclude *.manifest *.exp \
	$(COMMON_INC) $(HDEPENDS_FILES) $(PRODTARGETS) $(TESTPRODTARGETS) \
	$(TESTSCRIPTS) $(TAPFILES) $(JUNITFILES)
ifdef RES
	@$(RM) *$(RES)
endif
```

The problem is that OBJ isn't defined and so it winds up running:
```bash
rm -f *.i * *.a *.out MakefileInclude *.manifest *.exp
```

`rm -f *` attempts to remove all files in a directory.

colbysparks replied:
If you define OBJ in this Makefile, it won't be deleted on clean but that it is a total hack

colbysparks replied:
Ok the issue was that in `zaberMotionSupport` I was included `$(TOP)/configure/RULES`, which includes `epics-base/configure/RULES`, AND I had defined `T_A`, so in the end I was pulling in `epics-base/configure/RULES_BUILD`, which defines `clean` this way. I copied this from the ADPointGrey support folder without understanding what it meant.

Changing this to include `$(TOP)/configure/RULES_DIRS`, or simply not including any rules and defining a custom clean target works fine.

colbysparks committed:
Makefile updates
- Fix clean target in zaberMotionSupport
- Add zaberMotionSupport to iocs depend dirs

colbysparks committed:
Adds comment on defining runtime lib search path for ioc install

colbysparks committed:
Improves ioc example and scripting
- Adds tcp example to motor.cmd.zaber, updates usb ports
- Defines gulpfile script fn for running example ioc

colbysparks committed:
Improve zaberMotionSupport configuration
- Moves support lib env vars to ./configure/CONFIG_SITE
- Adds comment explaining why motorZaberMotion config site is included in ioc build

kmpeters commented:
I'm able to build the module now.  I'm building it as a standalone module, not as a submodule of motor.  I created a RELEASE.local file that looks like defines the path to `MOTOR`:

```
MOTOR=/home/username/epics/local/zaberReview/synApps_6_3/support/motor
-include $(MOTOR)/modules/RELEASE.$(EPICS_HOST_ARCH).local
# path to motorAcsMotion is needed to build the IOC inside motorZaberMotion, but outside motor
#!MOTOR_ZABER_MOTION=
```

Creating a RELEASE.local file or editing a top-level RELEASE file is normal.  I expect to do it for every EPICS module.

I also needed to correct the path to `MOTOR_ZABER_MOTION` in **motorZaberMotion/zaberMotionSupport/configure/RELEASE**, which is not something I normally expect to do.

My expectation is that the the libraries that are installed in the zaberMotionSupport/lib directory are instead installed to the motorZaberMotion/lib directory, instead of the libraries residing in two locations:

```
❯ ls -l lib/linux-x86_64 zaberMotionSupport/lib/linux_x64
lib/linux-x86_64:
total 10284
-r--r--r-- 1 username group 6996230 Nov 15 15:41 libZaberMotion.a
-r-xr-xr-x 1 username group 3480912 Nov 15 15:41 libZaberMotion.so

zaberMotionSupport/lib/linux_x64:
total 41912
-rw-rw-r-- 1 username group 12728208 Nov 15 09:22 libzaber-motion-core.so.7.2.0
lrwxrwxrwx 1 username group       22 Nov 15 09:22 libzaber-motion.so -> libzaber-motion.so.7.2
lrwxrwxrwx 1 username group       24 Nov 15 09:22 libzaber-motion.so.7.2 -> libzaber-motion.so.7.2.0
-rwxrwxr-x 1 username group 30003368 Nov 15 09:22 libzaber-motion.so.7.2.0
```

I think areaDetector drivers are much better examples of how to do this, since many of them include vendor libraries:

https://github.com/areaDetector/ADSpinnaker

colbysparks committed:
More changes to build configuration files
- Removes definition of T_A in support Makefile
- Replaces RULES_DIR with RULES in support Makefile
- Adds gulpfile fn for cleaning motor build directories

colbysparks commented on configure/RELEASE:
```diff
@@ -1,4 +1,5 @@
 # RELEASE - Location of external support modules
+MOTOR_ZABER_MOTION = $(MOTOR)/modules/motorZaberMotion
```
> I also needed to correct the path to MOTOR_ZABER_MOTION in motorZaberMotion/zaberMotionSupport/configure/RELEASE, which is not something I normally expect to do

Does it make sense to define `MOTOR_ZABER_MOTION` in `configure/RELEASE`, and give the user the option to override that in their own custom RELEASE.local file? It's definitely not an external support module, but if users are used to modifying RELEASE files then this might be the right place for it.

colbysparks replied:
Ok I just tested building this as standalone and as a motor submodule and it seems to be working. Support libs are being installed in the correct place and standalone configuration is as simple as creating a `RELEASE.local` file.

colbysparks committed:
Fix standalone build installation
- Update lib install location for standalone build
- Add comment on adding MOTOR_ZABER_MOTION path to RELEASE.local

colbysparks committed:
Fix standalone build installation
- Update lib install location for standalone build
- Add comment on adding MOTOR_ZABER_MOTION path to RELEASE.local

kmpeters committed:
Use a relative path to the include files to simplify the build configuration

kmpeters committed:
Added example CONFIG_SITE.local and RELEASE.local files

kmpeters committed:
Corrected path to include files

kmpeters commented:
I was able to further simplify the build configuration using a relative path to the include files.  I pushed my changes to the review-branch-build-improvements branch:

https://github.com/colbysparks/motorZaberMotion/compare/review-branch...review-branch-build-improvements?expand=1

I tested the build as a standalone module by copying the EXAMPLE_RELEASE.local and defining only the path to `MOTOR` and as a submodule of motor by adding the following line to `motor/modules/Makefile`:

```diff
diff --git a/modules/Makefile b/modules/Makefile
index 13b5cdc5..7de8cf0b 100644
--- a/modules/Makefile
+++ b/modules/Makefile
@@ -41,6 +41,7 @@ endif
 SUBMODULES += motorSmarAct
 SUBMODULES += motorSmartMotor
 SUBMODULES += motorThorLabs
+SUBMODULES += motorZaberMotion
 
 # Allow sites to add extra submodules
 -include Makefile.local
```

colbysparks commented on zaberMotionSupport/Makefile:
```diff
@@ -0,0 +1,17 @@
+$(error "motorZaberMotion: Unsupported architecture: $(EPICS_HOST_ARCH)")
+endif
+
+LIB_INSTALLS += $(wildcard ../ZaberMotionCppSupport/lib/$(ZML_HOST_ARCH)/*)
```
﻿>I was able to further simplify the build configuration using a relative path to the include files. I pushed my changes to the review-branch-build-improvements branch:
https://github.com/colbysparks/motorZaberMotion/compare/review-branch...review-branch-build-improvements?expand=1

Apparently there were a lot of cascading effects from my not having specified the `LIB_INSTALLS` path correctly. Now that this has been updated, the support libs always get installed to the correct location, so there is no longer an need to specify `-rpath` nor add any extra `-L` paths.

Thanks for figuring this out! When I saw other repos using `../` to install support libs I thought they might have been wrong, but it turns out that this path needs to be relative to the Makefile in `./O.linux-<arch>`, not relative to this file path..

I just tested your changes on `review-branch-build-improvements` on my linux machine and everything works, but I'm getting a runtime linking error running the example ioc on macOS. I'll commit these changes once I figure it out.

kmpeters replied:
I'm sorry I don't have an OS X machine I can use for testing.

colbysparks replied:
All good. I just discovered that the build system breaks symlinks when it copies the libs from the support package folder. I'll have to make some changes so that it doesn't just add three copies of the same library to the target lib install folder.

That said, even if `libzaber-motion.dylib` is a copy of the versioned lib it points to, runtime linking should still work on macOS so I doubt that's the issue I'm seeing.

<img width="1084" alt="image" src="https://github.com/user-attachments/assets/284dce81-9a62-4906-9a34-0153a5425e7d">

<img width="816" alt="image" src="https://github.com/user-attachments/assets/2c55c78e-6354-46b1-a40d-eb1c80c8c8ed">

colbysparks replied:
I haven't managed to find a good solution to the symlink copying issue in the EPICS build system, nor on the gulpfile scripting side of things. I might just have to make changes to how the zml support package is built, but that's a conversation I need to have with my team. I am fine leaving this for now but will make a note to fix some time in the short term.

As for the macOS problem, it's very likely that I could have missed something, but it seems like there may be an oversight in the EPICS build system where rpaths are not included for shared libs which need them to be specified. When linking the example ioc on linux (and `LINKER_USE_RPATH=YES`), the build system system will add rpaths for all the lib folders, including the folder which contains the support libs:
```
-Wl,-rpath,/home/ubuntu/EPICS/support/motor/lib/linux-aarch64
```
But for darwin builds no rpaths are passed to the linker, even with `LINKER_USE_RPATH=YES`, so it looks like a macOS user may have to manually add the rpath or use a command line tool for specifying an rpath in an existing executable.

kmpeters replied:
The "-rpath" option for Linux comes from this config file in EPICS base:

https://github.com/epics-base/epics-base/blob/b7cc33c3c9c1a6cc14290b9a558e97bd89171e80/configure/os/CONFIG.Common.linuxCommon#L26-L27

There isn't any reference to "rpath" in any of the darwin config files:

```
❯ pwd
/home/username/epics/local/zaberReview/base-7.0.8.1/configure/os
❯ grep rpath * | grep -i darwin
❯ 
```

I don't know if it is an oversight or not.  The EPICS base developers have a mailing list where they answer questions like this:

https://epics.anl.gov/core-talk/

colbysparks replied:
>The "-rpath" option for Linux comes from this config file in EPICS base:
https://github.com/epics-base/epics-base/blob/b7cc33c3c9c1a6cc14290b9a558e97bd89171e80/configure/os/CONFIG.Common.linuxCommon#L26-L27

I ended up finding this yesterday shortly after making that comment. Adding the following to the example IOC's `CONFIG_SITE` does fix the build issue with darwin:

```
PRODDIR_RPATH_LDFLAGS_YES += $(PROD_DEPLIB_DIRS:%=-Wl,-rpath,%)
PRODDIR_LDFLAGS += $(PRODDIR_RPATH_LDFLAGS_$(LINKER_USE_RPATH))
```

colbysparks replied:
I just emailed the core-talk mailing list, so I will wait to hear back from them on whether or not the above is an appropriate solution for darwin builds

colbysparks replied:
I heard back from the core talk forum and apparently my solution is fine, so I went and updated the build config and documentation for darwin users. I've also implemented `asynMotorAxis::setPosition`, and made a few other changes to improve error handling and reporting. 

I am starting to feel like this is ready for release. If you have any other comments or concerns, please let me know. Otherwise, thanks so much for taking the time to do this review. Your help has been invaluable.

colbysparks committed:
Merges changes from review-branch-build-improvements

colbysparks committed:
Updates build script and documentation
- Adds shared lib rpath for darwin PROD builds
- Updates README.md with to include instructions for darwin builds
- Adds documentation for setPosition

colbysparks committed:
Remove iocs history file and add to .gitignore

colbysparks committed:
Update ZML CPP Support version

colbysparks committed:
Merge branch 'dev' into review-branch

colbysparks committed:
Remove old so files

colbysparks committed:
Merge branch 'dev' into review-branch

colbysparks committed:
Replace software with contact email address in README.md

colbysparks committed:
Implements setPosition

colbysparks committed:
Improves error handling
- Wraps all calls to ZML api in try-catch blocks
- Adds newline to zaber utils handleException messages
- zaberController constructor error "Connection failed" -> "Initialization failed"

colbysparks committed:
Wraps controller report in try/catch

colbysparks committed:
Removes EPICS ci config files

colbysparks committed:
Update README to point new users to build instructions in SETUP
